samples: tflite_ethosu: support numaker_m55m1

ccli8 opened the pull request:
This adds Nuvoton's M55M1 support into the `tflm_ethosu` sample.
This sample is also refined for being able of adding other target support, including:

1. Fix undefined `_fini` symbol on invoking `exit()` for newlib libc 
1. Fix unsafe thread stack allocation. Use canonical K_THREAD_STACK_DEFINE or friends instead of raw buffer
1. Refactor Vela-compiled models which must match target Ethos-U RTL config
1. Add non-Vela-compiled model for evaluation of no-Ethos-U platform
1. Add support for **numaker_m55m1** board

**NOTE**: This PR is continuation of #88941.

ccli8 committed:
samples: tflite_ethosu: fix unsafe thread stack allocation

It is unsafe to pass raw buffer as thread stack. Instead, use canonical
K_THREAD_STACK_DEFINE or friends to define thread stack.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 commented on samples/modules/tflite-micro/tflm_ethosu/src/npu_cache.c:
```diff
@@ -0,0 +1,31 @@
+
+#ifndef CONFIG_NOCACHE_MEMORY
+
+void ethosu_flush_dcache(uint32_t *p, size_t bytes)
```
Updated, thanks for informing of this.

ccli8 commented on samples/modules/tflite-micro/tflm_ethosu/src/models/keyword_spotting_cnn_small_int8_ethos_u55_256/model.h:
Originally, there has been one such [model source](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/modules/tflite-micro/tflm_ethosu/src/models/keyword_spotting_cnn_small_int8/model.h) of that size. However, it is vela-compiled for ethos-u55-128. I have one idea: keep just one model source but it changes to non-vela-compiled version, so that all targets with or without ethos-u hardware can run it. Then for targets with ethos-u hardware, users need to vela-compile model which must meet their ethos-u hardware RTL config additionally. This scheme doesn't add source size due to model source and adds feasibility of supporting other targets with differing ethos-u hardware.

ccli8 replied:
Following https://github.com/zephyrproject-rtos/zephyr/pull/93902#issuecomment-3250081392, remove all model blobs and add guide on how to generate Vela-compiled model into build.

khoa-nguyen-18 reviewed (dismissed):
Confirmed the sample app runs successfully on the Renesas MCU. Thanks

ccli8 commented:
> I would prefer to have one default model that just work out of box for MPS3 (the original one). But if that is not an option could we at least fail to compile if we have an empty model?

Failing the compile causes CI failure.  We need to make one resolution among keep one default model, build-only, or other means. @johan-alfven-arm @JarmouniA

johan-alfven-arm commented:
> > I would prefer to have one default model that just work out of box for MPS3 (the original one). But if that is not an option could we at least fail to compile if we have an empty model?
> 
> Failing the compile causes CI failure. We need to make one resolution among keep one default model, build-only, or other means. @johan-alfven-arm @JarmouniA

I vote keep the default model. @JarmouniA what do you say?

ccli8 commented:
> Seems reasonable, but should be clarified in the sample doc, so that there are no future attempts to add other ones or change the default to suit a specific platform.

I will keep default model originally for MPS3/MPS4, and add above into readme.

johan-alfven-arm reviewed (commented):
Ahh, the default example will actually only run on Ethos-U55 (MPS3). So you can remove MPS4

JarmouniA reviewed (commented):
SonarQube issue needs to be fixed https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=93902&id=zephyrproject-rtos_zephyr&open=AZheU6KdVNUBhAn6e_pl

ccli8 commented:
> Ahh, the default example will actually only run on Ethos-U55 (MPS3). So you can remove MPS4

Update note:
1. Keep default model blob for MPS3 target
2. All other targets are build-only with warning (pass CI).

ccli8 commented:
Update notes:
1. Make `CONFIG_TAINT_BLOBS_TFLM_ETHOSU` depend on `CONFIG_ETHOS_U_ARM` or `CONFIG_ETHOS_U_NUMAKER`
2. Remove warning message for non-MSP3 target to pass CI

kartben commented:
I would like @stephanosio to chime in regarding the proposed libc change

stephanosio commented on lib/libc/newlib/libc-hooks.c:
```diff
@@ -307,6 +307,20 @@ __weak void _exit(int status)
 	}
 }
 
+/* Address undefined _fini for the exit call
+ *
+ * With gcc command-line options -nostartfiles or -nostdlib, the gcc
+ * built-in object files crti.o/crtn.o which implement _init/_fini
+ * won't get linked in automatically and will cause undefined reference
+ * to _fini error when exit is invoked.
+ *
+ * This is fixed by providing one dummy _fini to let linker pass.
+ */
+__weak void _fini(void)
+{
+	/* This function is usually overridden and thus it's kept empty. */
+}
```
AFAICS, this should not be necessary for Zephyr SDK. Are you using newlib from some other toolchain?

ccli8 commented:
Update notes:
1. Rebase on `main` branch
2. Drop the `__fini` override commit

kartben commented on samples/modules/tflite-micro/tflm_ethosu/README.rst:
```diff
@@ -19,6 +19,65 @@ where the operators supported by Ethos-U have been replaced by an Ethos-U custom
+
+   Assuming target Ethos-U is U55 and 128 MAC:
+
+.. code-block:: console
+
+    $ vela cnn_s_quantized.tflite \
+    --output-dir . \
+    --accelerator-config ethos-u55-128
```
```suggestion
   .. code-block:: console

       $ vela cnn_s_quantized.tflite \
       --output-dir . \
       --accelerator-config ethos-u55-128
```

kartben commented on samples/modules/tflite-micro/tflm_ethosu/README.rst:
```diff
@@ -19,6 +19,65 @@ where the operators supported by Ethos-U have been replaced by an Ethos-U custom
+   ``testing_input/input/0.npy`` and ``testing_output/0.npy`` have 128-byte header.
+   They must be removed for integration with this sample.
+
+.. code-block:: console
+
+    $ dd if=testing_input/input/0.npy of=testing_input/input/0_no-header.npy bs=1 skip=128
+    $ dd if=testing_output/identity/0.npy of=testing_output/identity/0_no-header.npy bs=1 skip=128
```
```suggestion
   .. code-block:: console

       $ dd if=testing_input/input/0.npy of=testing_input/input/0_no-header.npy bs=1 skip=128
       $ dd if=testing_output/identity/0.npy of=testing_output/identity/0_no-header.npy bs=1 skip=128
```

kartben commented on samples/modules/tflite-micro/tflm_ethosu/README.rst:
```diff
@@ -19,6 +19,65 @@ where the operators supported by Ethos-U have been replaced by an Ethos-U custom
+
+4. Converting to C array
+
+.. code-block:: console
+
+    $ xxd -c 16 -i cnn_s_quantized.tflite cnn_s_quantized.tflite.h
+    $ xxd -c 16 -i cnn_s_quantized_vela.tflite cnn_s_quantized_vela.tflite.h
+    $ xxd -c 16 -i testing_input/input/0_no-header.npy testing_input/input/0_no-header.npy.h
+    $ xxd -c 16 -i testing_output/identity/0_no-header.npy testing_output/identity/0_no-header.npy.h
```
```suggestion
   .. code-block:: console

       $ xxd -c 16 -i cnn_s_quantized.tflite cnn_s_quantized.tflite.h
       $ xxd -c 16 -i cnn_s_quantized_vela.tflite cnn_s_quantized_vela.tflite.h
       $ xxd -c 16 -i testing_input/input/0_no-header.npy testing_input/input/0_no-header.npy.h
       $ xxd -c 16 -i testing_output/identity/0_no-header.npy testing_output/identity/0_no-header.npy.h
```

kartben commented on samples/modules/tflite-micro/tflm_ethosu/README.rst:
```diff
@@ -19,6 +19,65 @@ where the operators supported by Ethos-U have been replaced by an Ethos-U custom
+   - testing_input/input/0_no-header.npy.h > input.h
+   - testing_output/identity/0_no-header.npy.h > output.h
+
+.. note:: To run non-Vela-compiled model (``CONFIG_TAINT_BLOBS_TFLM_ETHOSU=n``),
+   synchronize ``cnn_s_quantized.tflite.h`` instead.
```
```suggestion
   .. note:: To run non-Vela-compiled model (``CONFIG_TAINT_BLOBS_TFLM_ETHOSU=n``),
      synchronize ``cnn_s_quantized.tflite.h`` instead.
```

ccli8 replied:
fixed spacing

johan-alfven-arm commented on samples/modules/tflite-micro/tflm_ethosu/src/models/keyword_spotting_cnn_small_int8/input.h:
```diff
@@ -4,7 +4,15 @@
+#endif
+
+INPUT_ATTR uint8_t inputData[] = {
+#if defined(CONFIG_SOC_SERIES_MPS3)
```
No need for input do depend on SOC.

ccli8 committed:
samples: tflite_ethosu: support numaker_m55m1

This supports nuvoton numaker_m55m1 board, including.

Besides, re-organize for new target and Ethos-U support, including:
1. Keep only one model originally for MPS3 target
2. Add guide to regenerate Vela-compiled model for targets supporting
   Ethos-U
3. Support running non-Vela-compiled model for targets supporting no
   Ethos-U

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>
